model/minsky.cc

```diff
@@ -713,6 +726,96 @@ namespace minsky
   }
+  
+  /*
+  For using GSL non-linear least square fitting routines
+  patterned on first Large Nonlinear Least Squares Example.
+  See online documentation at https://www.gnu.org/software/gsl/doc/html/nls.html
+  */
+      
+namespace
+{	
+
+   int residualFunc(const gsl_vector * x, void *params,
+          gsl_vector * f)
+   {
+	if (params==NULL) return GSL_EBADFUNC;
+	try { 	  
+		
+	      auto pars=static_cast<OptParams*>(params);  
+	      
+          int err = GSL_SUCCESS;
+          auto n = ValueVector::stockVars.size();
+          vector<double> xx(n);
+          
+          memcpy(xx.data(),reinterpret_cast<const vector<double>*>(x),n);
+          
+          err=gsl_odeiv2_driver_apply(pars->d->driver, pars->t,numeric_limits<double>::max(),&xx[0]);
+	      
+		  // define residual vector function to be optimized 
+          for (size_t i = 0; i < n; i++)
+              gsl_vector_set (f, i, xx[i] - pars->xi[i]);  
+     }
+     catch  (std::exception& e) {
+		((Minsky*)params)->threadErrMsg=e.what();
+        return GSL_EBADFUNC; 
+	 }   
+	 return  GSL_SUCCESS;      
+                        
+   }
+   
+   static void errHandlerOpt(const char* reason, const char* file, int line, int gsl_errno) {
+         throw error("gsl: %s:%d: %s",file,line,reason);
+   } 
+}      
+
+  int Minsky::optimize(double xf[], OptParams& optPars)
+  {
+    	
+      gsl_set_error_handler(errHandlerOpt);	
+      const gsl_multifit_nlinear_type *T = gsl_multifit_nlinear_trust;
+      gsl_multifit_nlinear_workspace *w;
+      gsl_multifit_nlinear_fdf fdf;
+      gsl_multifit_nlinear_parameters fdf_params = gsl_multifit_nlinear_default_parameters();	
+  	
+      const size_t n = stockVars.size();   
+      const size_t p = n; 
+      
+      gsl_vector *f = gsl_vector_alloc(n);
+      gsl_vector_view x = gsl_vector_view_array (xf, p);    
+      int status, info;
+      
+      const double xtol = 1e-12;
+      const double gtol = 1e-12;
+      const double ftol = 0.0;
+      
+      /* define the function to be minimized */
+      fdf.f = residualFunc;
+      fdf.df = NULL; //func_df;   /* set to NULL for finite-difference Jacobian */
+      fdf.fvv = NULL;     /* not using geodesic acceleration */
+      fdf.n = n;
+      fdf.p = p;
+      fdf.params = &optPars; 
+      
+      /* allocate workspace with default parameters */
+      w = gsl_multifit_nlinear_alloc (T, &fdf_params, n, p);
+      
+      /* initialize solver with starting point and weights */
+      gsl_multifit_nlinear_init (&x.vector, &fdf,w);
+      
+      /* compute initial cost function */
+      f = gsl_multifit_nlinear_residual(w);
+      
+      /* solve the system with a maximum of 200 iterations */
+      status = gsl_multifit_nlinear_driver(200, xtol, gtol, ftol,
+                                           NULL, NULL, &info, w);
+                                           
+      memcpy(xf,reinterpret_cast<const vector<double>*>(w->x),n);                                     
```

Comment:
The cast is wrong. Is w->x a simple array of doubles?

Reply:
I have replaced it with memcpy(xf,w->x,n) as inside residualFunc, but am dubious about the size of w->x too.